### PR TITLE
refactor: unify client test scenarios 

### PR DESCRIPTION
**Pull request by teamdandelion**

### CODE REVIEW

**Comment by teamdandelion** [comment]
* **#1288** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1288?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#1293](https://github.com/Mirascope/mirascope/pull/1293) <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1293?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#1300](https://github.com/Mirascope/mirascope/pull/1300) <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1300?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1291** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1291?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1287** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1287?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1292** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1292?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1289** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1289?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1289?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1290** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1290?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `v2`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python/tests/llm/clients/anthropic/test_anthropic_client_call.py
Comment (willbakst) [inline python/tests/llm/clients/anthropic/test_anthropic_client_call.py:1]: nit: now that this file only has a single test, I feel like a single `test_anthropic_client.py` with `test_call` and `test_stream` might make more sense?

File: python/tests/llm/clients/anthropic/test_anthropic_client_call.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+)
+from tests import utils
+from tests.llm.clients.conftest import CLIENT_SCENARIOS
+
+
+@pytest.mark.parametrize("scenario_fixture", CLIENT_SCENARIOS)
+@pytest.mark.vcr()
+def test_call(
+    anthropic_client: llm.AnthropicClient,
+    scenario_fixture: str,
```
Comment (willbakst) [inline python/tests/llm/clients/anthropic/test_anthropic_client_call.py]: nit: consider naming this something like `scenario_name` or `scenario_key`? Cause I read `scenario_fixture` as the fixture (value) itself, but you have to do `request.getfixturevalue` on the key to get the actual value (as you do below)

Perhaps something like `scenario_id` or `scenario_fixture_id`? Could then update `CLIENT_SCENARIOS` to `CLIENT_SCENARIO_IDS` or `CLIENT_SCENARIO_FIXTURE_IDS`?

File: python/tests/llm/clients/google/test_google_client.py [hunk 2]
```diff
@@ -27,74 +27,7 @@ def test_custom_base_url() -> None:
...
-            ],
-            "content": [llm.Text(text="Hello world\n")],
-            "texts": [llm.Text(text="Hello world\n")],
-            "thoughts": [],
-            "tool_calls": [],
-        }
-    )
-
-
+# Included in test_google_client as this triggers a google-specific edge case for coverage.
```
Comment (willbakst) [inline python/tests/llm/clients/google/test_google_client.py:437]: nit: In line with my previous comment, perhaps it would make the most sense to have `test_{provider}_client.py` as the core test file and then optionally have `test_{provider}_client_edge_cases.py` if we want to separate them?

File: python/tests/llm/clients/conftest.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        elif location == "SF":
+            return "The weather in SF is overcast and 64°F"
+        else:
+            return "Unknown city " + location
+
+    return get_weather
+
+
+@pytest.fixture
+def simple_message() -> CallKwargs:
```
Comment (willbakst) [inline python/tests/llm/clients/conftest.py]: nit: I feel like we should put `_scenario` for each of these? e.g. `simple_message_scenario`?

File: python/tests/llm/clients/conftest.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        "messages": [
+            llm.messages.user("What is the weather in SF and NYC?"),
+            llm.messages.assistant([tool_call_sf, tool_call_nyc]),
+            llm.messages.user([tool_output_sf, tool_output_nyc]),
+        ],
+        "tools": [weather_tool],
+    }
+
+
+CLIENT_SCENARIOS = [
```
Comment (willbakst) [inline python/tests/llm/clients/conftest.py]: nit: I wonder if, given my other comment re: `_scenario`, we could use something like `importlib` or `inspect` or `ast` to dynamically generate the list of scenarios by snagging all function names that end with `_scenarios` so that we don't have to worry when adding a new scenario that we've added it to this list?

File: python/tests/llm/clients/anthropic/test_anthropic_client_call.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+- San Francisco (SF) is overcast with a temperature of 64°F
+- New York City (NYC) is sunny with a temperature of 72°F\
+"""
+                                )
+                            ]
+                        ),
+                    ],
+                },
+            }
+        )[scenario_fixture]
```
Comment (willbakst) [inline python/tests/llm/clients/anthropic/test_anthropic_client_call.py]: Does this have to be written like this for `inline-snapshot` to work? Or could we do something like this?

```python
expected = snapshot({ ... })
assert utils.response_snapshot_dict(response) == expected[scenario_id]
```

Comment (teamdandelion) [reply python/tests/llm/clients/anthropic/test_anthropic_client_call.py]: Changed it.

File: python/tests/llm/clients/conftest.py (hunk 4 above)
Comment (teamdandelion) [reply python/tests/llm/clients/conftest.py]: Nice suggestion, done

File: python/tests/llm/clients/anthropic/test_anthropic_client_call.py (hunk 1 above)
Comment (teamdandelion) [reply python/tests/llm/clients/anthropic/test_anthropic_client_call.py]: Renamed to scenario_id

File: python/tests/llm/clients/openai/test_openai_client.py [hunk 6]
```diff
@@ -1,1470 +1,908 @@
...
+                    "title": "THE BROKEN EMPIRE TRILOGY",
+                    "author": "MARK LAWRENCE",
+                    "vibe": "HEART-WRENCHING BRUTALITY AND GUT-PUNCHING TRAGEDY!",
+                },
+            },
+            "gpt-4:strict": {
+                "exception_type": "BadRequestError",
+                "exception_message": "Error code: 400 - {'error': {'message': \"Invalid parameter: 'response_format' of type 'json_schema' is not supported with this model. Learn more about supported models at the Structured Outputs guide: https://platform.openai.com/docs/guides/structured-outputs\", 'type': 'invalid_request_error', 'param': None, 'code': None}}",
+                "status_code": 400,
+            },
```
Comment (teamdandelion) [inline python/tests/llm/clients/openai/test_openai_client.py:1114]: Expected, though maybe we should have a generic Mirascope error in this case.

File: python/tests/llm/clients/openai/test_openai_client.py [hunk 7]
```diff
@@ -1,1470 +1,908 @@
...
+                "system_message": """\
+You are a depressive LLM that only recommends sad books.
+Pretty please output a book recommendation in JSON form.
+It should have the format {title: str, author: str, vibe: str}.
+Be super vibe-y with the vibe and make sure EVERYTHING IS CAPS to convey
+the STRENGTH OF YOUR RECOMMENDATION!\
+""",
+                "model_dump": {
+                    "title": "The Song of Achilles",
+                    "author": "Madeline Miller",
```
Comment (teamdandelion) [inline python/tests/llm/clients/openai/test_openai_client.py]: Didn't follow directions well in this case (in terms of capitalizing everything). Great book though

File: python/tests/llm/clients/openai/test_openai_client.py [hunk 8]
```diff
@@ -1,786 +1,822 @@
...
-    """Test handling of assistant messages with multiple text content parts.
+    """Test final formatted output for every supported formatting mode.
 
-    This test is added for coverage. It's slightly contrived as OpenAI assistant messages
-    only ever have a single string content, however it could come up if one is re-playing
-    message history generated by a different LLM into the OpenAI APIs.
+    This test uses the basic Book formatting. Expected qualities of the output:
+    - should have title, author, vibe as string
+    - vibe should be mysterious, euphoric, intruiging, or soul-searching (per annotation)
+    - vibe should always have an exclamation point (per class docstring)
```
Comment (teamdandelion) [inline python/tests/llm/clients/openai/test_openai_client.py:747]: It mostly does not respect this property.

File: python/tests/llm/clients/openai/cassettes/test_every_structured_mode_basic[gpt-4o-strict-or-json].yaml [hunk 9]
```diff
@@ -0,0 +1,109 @@
+interactions:
+- request:
+    body: '{"messages":[{"role":"user","content":"Recommend a single fantasy book."}],"model":"gpt-4o","response_format":{"type":"json_schema","json_schema":{"name":"Book","schema":{"description":"A
+      book recommendation with metadata. ALWAYS add an exclamation point to the vibe!","properties":{"title":{"title":"Title","type":"string"},"author":{"title":"Author","type":"string"},"vibe":{"description":"Should
+      be one of mysterious, euphoric, intruiging, or soul_searching","title":"Vibe","type":"string"}},"required":["title","author","vibe"],"title":"Book","type":"object","additionalProperties":false},"strict":true,"description":"A
```
Comment (teamdandelion) [inline python/tests/llm/clients/openai/cassettes/test_every_structured_mode_basic[gpt-4o-strict-or-json].yaml:5]: Note: the descriptin (Always an excalamtion point to the vibe!) is getting added twice in this case, via json_schema.schema.description and via json_schema.description. Not due to this PR, but may want to change behavior here to deduplicate.

File: python/mirascope/llm/clients/openai/_utils.py [hunk 10]
```diff
@@ -222,6 +222,7 @@ def prepare_openai_request(
     shared_openai_types.ResponseFormatJSONObject
```
Comment (willbakst) [inline python/mirascope/llm/clients/openai/_utils.py:222]: <p class='graphite__hidden'><i>[Re: line +218]</i></p>
NOTE: We should discuss potentially updating these preparation methods to return something like `tuple[Sequence[Message], OpenAIKwargs]` so that we don't have to keep updating the return type and unpacking handling and can instead just receive `input_messages` and `kwargs`.

We would also then avoid the `tool_choice = NotGiven()` since we can instead just have `kwargs` without `tool_choice` and set `kwargs["tool_choice"]` if the conditional wants that. Then we can just do `client.chat.completions.create(**kwargs)` in the actual request.

No expectations around doing this here, just noting this as something we might want to do in a following PR that could clean things up a bit long term.
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1289?utm_source=unchanged-line-comment">Graphite</a>.</p>

File: python/tests/llm/clients/conftest.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+from typing import Annotated
+from typing_extensions import TypedDict
+
+import pytest
+from pydantic import BaseModel, Field
+
+from mirascope import llm
+
+
+class CallKwargs(TypedDict, total=False):
```
Comment (willbakst) [inline python/tests/llm/clients/conftest.py:15]: NOTE: I don't know if this would be a good idea, but we could move this into the `clients` module and use `**kwargs: Unpack[CallKwargs]` in each of the client calls to unify where we define the client call kwargs. I don't know the implications of this on downstream docstrings, so food for thought.

File: python/tests/llm/clients/openai/test_openai_client.py [hunk 12]
```diff
@@ -1269,202 +918,228 @@ class SimpleBook(BaseModel):
...
+@pytest.mark.parametrize("format_mode", FORMATTING_MODES)
 @pytest.mark.parametrize(
-    "mode", ["strict", "strict-or-tool", "strict-or-json", "tool", "json"]
-)
+    "model_id", ["gpt-4o", "gpt-4"]
+)  # gpt-4o has native strict and json support; gpt-4 does not.
 @pytest.mark.vcr()
-def test_structured_stream_supported_modes(
-    openai_client: llm.OpenAIClient, mode: llm.formatting.FormattingMode
+def test_every_structured_mode_with_instructions_and_system_message(
```
Comment (willbakst) [inline python/tests/llm/clients/openai/test_openai_client.py:966]: this feels like a lot of duplicated code with the other structured output all modes test. why not just have these as fixtures like the rest and combine the tests?

File: python/tests/llm/clients/openai/test_openai_client.py (hunk 7 above)
Comment (willbakst) [reply python/tests/llm/clients/openai/test_openai_client.py]: It's sad when it doesn't follow directions. Bad model.

File: python/tests/llm/clients/google/test_google_client.py [hunk 13]
```diff
@@ -6,7 +6,414 @@
...
+                            )
+                        ]
+                    ),
+                ],
+                "consumed": True,
+                "n_chunks": 5,
+            },
+        }
+    )
+    assert utils.stream_response_snapshot_dict(stream_response) == expected[scenario_id]
```
Comment (willbakst) [inline python/tests/llm/clients/google/test_google_client.py:416]: nit: please add a newline before all of these assertions across the client tests, I find that easier to read

File: python/tests/llm/clients/openai/test_openai_client.py (hunk 6 above)
Comment (willbakst) [reply python/tests/llm/clients/openai/test_openai_client.py:1114]: Yeah this goes back to my talk about unifying this check in some way, in which case the unified check could raise a Mirascope error mentioning the issue.

In a perfect world, users would never have to think about the underlying SDKs or their errors since Mirascope would handle them.

File: python/tests/llm/clients/openai/test_openai_client.py (hunk 8 above)
Comment (willbakst) [reply python/tests/llm/clients/openai/test_openai_client.py:747]: Probably just a prompting issue. Maybe something like `"""the vibe MUST have the form "{vibe}!" with an exclamation mark included"""`

File: python/tests/llm/clients/openai/cassettes/test_every_structured_mode_basic[gpt-4o-strict-or-json].yaml (hunk 9 above)
Comment (willbakst) [reply python/tests/llm/clients/openai/cassettes/test_every_structured_mode_basic[gpt-4o-strict-or-json].yaml:5]: The schema also includes "title" as a field even when the title is just the capitalized version of the field, in which case it's a waste of tokens. We should really only ever add title when it's different from the field name.

In a following PR we should definitely remove any duplication that is just a waste of tokens.

File: python/tests/llm/clients/conftest.py (hunk 11 above)
Comment (teamdandelion) [reply python/tests/llm/clients/conftest.py:15]: Ack, will keep it as food for thought.

File: python/tests/llm/clients/openai/test_openai_client.py (hunk 6 above)
Comment (teamdandelion) [reply python/tests/llm/clients/openai/test_openai_client.py:1114]: Discussed offline, we'll do this later.

File: python/tests/llm/clients/openai/test_openai_client.py (hunk 12 above)
Comment (teamdandelion) [reply python/tests/llm/clients/openai/test_openai_client.py:966]: Updated for all 3 providers in #1293

File: python/tests/llm/clients/google/test_google_client.py (hunk 13 above)
Comment (teamdandelion) [reply python/tests/llm/clients/google/test_google_client.py:416]: Done for all providers in #1293

File: python/mirascope/llm/clients/openai/_utils.py (hunk 10 above)
Comment (teamdandelion) [reply python/mirascope/llm/clients/openai/_utils.py:222]: Discussed, this is a nice idea for a followon, may add to this stack.

File: python/tests/llm/clients/openai/cassettes/test_every_structured_mode_basic[gpt-4o-strict-or-json].yaml (hunk 9 above)
Comment (teamdandelion) [reply python/tests/llm/clients/openai/cassettes/test_every_structured_mode_basic[gpt-4o-strict-or-json].yaml:5]: Ack, adding to my notes for a follow up.

File: python/tests/llm/clients/openai/test_openai_client.py (hunk 8 above)
Comment (teamdandelion) [reply python/tests/llm/clients/openai/test_openai_client.py:747]: Ack, will play around with this when I'm doing future cleanup that will churn snapshots.

**Review by willbakst — APPROVED** [review]

**Comment by teamdandelion** [comment]
### Merge activity

* **Sep 11, 6:14 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1289).
* **Sep 11, 6:16 AM UTC**: [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1289) rebased this pull request as part of a merge.
* **Sep 11, 6:17 AM UTC**: @teamdandelion merged this pull request with [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1289).
